### PR TITLE
Make the SHAPE_TOO_COMPLEX performance warning more actionable

### PR DESCRIPTION
**Pull request by casperisfine**
As suggested by @mame, we should try to help users fix the issues without having to lookup the meaning of the warning.

### CODE REVIEW

File: shape.c [hunk 1]
```diff
@@ -733,9 +733,10 @@ rb_shape_get_next(rb_shape_t *shape, VALUE obj, ID id)
                 if (RCLASS_EXT(klass)->variation_count >= SHAPE_MAX_VARIATIONS) {
                     rb_category_warn(
                         RB_WARN_CATEGORY_PERFORMANCE,
-                        "Maximum shapes variations (%d) reached by %"PRIsVALUE", instance variables accesses will be slower.",
-                        SHAPE_MAX_VARIATIONS,
-                        rb_class_path(klass)
+                        "The class %"PRIsVALUE" reached %d shape variations, instance variables accesses will be slower amd memory usage increased.\n"
```
Comment (mame) [inline shape.c:736]:
```suggestion
                        "The class %"PRIsVALUE" reached %d shape variations, instance variables accesses will be slower and memory usage increased.\n"
```

File: test/ruby/test_object.rb [hunk 2]
```diff
@@ -458,7 +458,7 @@ def test_check_to_integer
   end
 
   def test_max_shape_variation_with_performance_warnings
-    assert_in_out_err([], <<-INPUT, %w(), /Maximum shapes variations \(8\) reached by Foo, instance variables accesses will be slower\.$/)
+    assert_in_out_err([], <<-INPUT, %w(), /The class Foo reached 8 shape variations, instance variables accesses will be slower amd memory usage increased/)
```
Comment (mame) [inline test/ruby/test_object.rb:461]:
```suggestion
    assert_in_out_err([], <<-INPUT, %w(), /The class Foo reached 8 shape variations, instance variables accesses will be slower and memory usage increased/)
```

**Review by mame — COMMENTED** [review]
😇

File: shape.c (hunk 1 above)
Comment (casperisfine) [reply shape.c]: Thanks, for my defense I'm not done drinking my ☕  😉

**Review by mame — APPROVED** [review]

File: shape.c [hunk 3]
```diff
@@ -733,9 +733,10 @@ rb_shape_get_next(rb_shape_t *shape, VALUE obj, ID id)
                 if (RCLASS_EXT(klass)->variation_count >= SHAPE_MAX_VARIATIONS) {
                     rb_category_warn(
                         RB_WARN_CATEGORY_PERFORMANCE,
-                        "Maximum shapes variations (%d) reached by %"PRIsVALUE", instance variables accesses will be slower.",
-                        SHAPE_MAX_VARIATIONS,
-                        rb_class_path(klass)
+                        "The class %"PRIsVALUE" reached %d shape variations, instance variables accesses will be slower and memory usage increased.\n"
+                        "It is recommended to define instance variable in a consistent order, for instance by eagerly defining them all in the `initialize` method.",
```
Comment (eregon) [inline shape.c:737]: Minor, but backquotes seem rare in Ruby exception messages, here I think `#initialize` would be better.

Comment (byroot) [reply shape.c:737]: https://github.com/ruby/ruby/pull/9280

**Pull request merged**